lib/AST/Decl.cpp

```diff
@@ -2092,9 +2092,39 @@ bool NominalTypeDecl::derivesProtocolConformance(ProtocolDecl *protocol) const {
     case KnownProtocolKind::BridgedNSError:
       return isObjC() && enumDecl->hasOnlyCasesWithoutAssociatedValues();
 
+    // Enums without associated values and enums with a raw type of String
+    // or Int can explicitly derive CodingKey conformance.
+    case KnownProtocolKind::CodingKey: {
+      Type rawType = enumDecl->getRawType();
+      if (rawType) {
+        auto parentDC = enumDecl->getDeclContext();
+        ASTContext &C = parentDC->getASTContext();
+
+        auto nominal = rawType->getAnyNominal();
+        return nominal == C.getStringDecl() || nominal == C.getIntDecl();
+      } else {
+        // Empty enums are allowed to conform as well.
+        return enumDecl->getAllElements().empty() ||
+               enumDecl->hasOnlyCasesWithoutAssociatedValues();
+      }
+    }
+
     default:
       return false;
     }
+  } else if (isa<StructDecl>(this) || isa<ClassDecl>(this)) {
```

Comment:
This violates LLVM style:

http://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

All of the code paths before the else if return. The else is not necessary. In fact, you could even invert the if condition and do an early return false if you want to.